addons/token-exchange/agent_mirrorpeer_controller.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2021 Red Hat OpenShift Data Foundation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package addons
+
+import (
+	"context"
+	"fmt"
+
+	replicationv1alpha1 "github.com/csi-addons/volume-replication-operator/api/v1alpha1"
+	multiclusterv1alpha1 "github.com/red-hat-storage/odf-multicluster-orchestrator/api/v1alpha1"
+	"github.com/red-hat-storage/odf-multicluster-orchestrator/controllers/common"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// MirrorPeerReconciler reconciles a MirrorPeer object
+type MirrorPeerReconciler struct {
+	client.Client
+	Scheme          *runtime.Scheme
+	HubKubeConfig   *rest.Config
+	SpokeKubeConfig *rest.Config
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *MirrorPeerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.Infof("In agent reconciler")
+	// Fetch MirrorPeer for given Request
+	var mirrorPeer multiclusterv1alpha1.MirrorPeer
+	err := r.Get(ctx, req.NamespacedName, &mirrorPeer)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Info("Could not find MirrorPeer. Ignoring since object must have been deleted")
+			return ctrl.Result{}, nil
+		}
+		klog.Error(err, "Failed to get MirrorPeer")
+		return ctrl.Result{}, err
+	}
+
+	err = r.createVolumeReplicationClass(ctx, &mirrorPeer)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create volumereplicationclass: %v", err)
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *MirrorPeerReconciler) createVolumeReplicationClass(ctx context.Context, mp *multiclusterv1alpha1.MirrorPeer) error {
+	params := make(map[string]string)
+	params[MirroringModeKey] = string(mp.Spec.Mode)
+	params[SchedulingIntervalKey] = mp.Spec.SchedulingInterval
+	params[ReplicationSecretNameKey] = mp.Spec.ReplicationSecretName
+	params[ReplicationSecretNamespaceKey] = StorageNamespace
```

Comment:
Let's make this namespace dynamic. Since we have MirrorPeer object, we know the name and namespace of storagecluster also.

Reply:
done